docs/DevGuide/Parallelization.md

```diff
@@ -564,6 +571,103 @@ calls on all array elements occurred in the same order. It is **undefined**
+race conditions, it is
+important that up-to-date checks are based on something that is
+independent of the order of mutation (like a `temporal_id`, and not
+like checking the size of a vector).
+
+To check an item, use the function
+`Parallel::mutable_cache_item_is_ready`, which returns a bool
+indicating whether the item is up to date.  If the item is up to date,
+then it can be used.  `Parallel::mutable_cache_item_is_ready` takes a
+lambda as an argument.  This lambda is passed a single argument: a const
+reference to the item being retrieved.  The lambda should determine
+whether the item is up to date. If so, it should return boost::none;
+if not, it should return a boost::optional<CkCallback> to a callback
+function that will be called on the next `Parallel::mutate` of that
+item. The callback will typically check again if the item is up to date
+and if so will execute some code that gets the item via `Parallel::get`.
+
+For the case of iterable actions, `Parallel::mutable_cache_item_is_ready`
+is typically called from the `is_ready` function of the iterable action,
+and the callback is `perform_algorithm()`.  In the example below, the
+vector is considered up to date if it is non-empty:
+
+\snippet Test_AlgorithmGlobalCache.cpp check_mutable_cache_item_is_ready
+
+Note that `Parallel::mutable_cache_item_is_ready` is called on a local
+core and does no parallel communication.
+
+### 2. Retrieving the item
+
+The item is retrieved using `Parallel::get` just like for constant items.
+For example, to retrieve the item `Tags::VectorOfDoubles`:
+\snippet Test_AlgorithmGlobalCache.cpp retrieve_mutable_cache_item
+
+Note that `Parallel::get` is called on a local core and does no
+parallel communication.
+
+Whereas we support getting *non-mutable* items in the GlobalCache from
+a databox via `db::get`, we intentionally do not support
+`db::get` of *mutable* items from a databox.  The reason is that mutable
+items should be retrieved only after a `Parallel::mutable_cache_item_is_ready`
+check, and being able to retrieve a mutable item from a databox makes it
+difficult to enforce that check, especially when automatically-executing
+compute items are considered.
+
+## 3. Modifying a mutable GlobalCache item
+
+To modify a mutable item, pass `Parallel::mutate` two template
+parameters: the tag to mutate, and a struct with an `apply` function
+that does the mutating. `Parallel::mutate` takes two arguments:
+a proxy to the GlobalCache, and a tuple that is passed into the
+mutator function.  For the following example,
+
+\snippet Test_AlgorithmGlobalCache.cpp mutate_global_cache_item
+
+the mutator function is defined as below:
+\snippet Test_AlgorithmGlobalCache.cpp mutate_global_cache_item_mutator
+
+`Parallel::mutate` broadcasts to every core, where it calls the
+mutator function and then calls all the callbacks that have been set
+on that core by `Parallel::mutable_cache_item_is_ready`.
+
```

Comment:
Note that the `Parallel::mutate` operation is guaranteed to be thread-safe without any further action by the developer